### ISSUE TITLE
Enable viewing BAM files in IGV

### ISSUE THREAD

**Issue by hackdna** [parklab/refinery-platform#1264]
Currently, the following error is displayed by IGV when trying to visualize a BAM file:
`Errors were encountered loading the session: http://test.stemcellcommons.org/media/file_store/15/d4/1_tophat_accepted_hits.bam &nbs;p org.broad.igv.exceptions.DataLoadException: An error occurred while accessing: http://test.stemcellcommons.org/media/file_store/15/d4/1_tophat_accepted_hits.bai Index file not found. Tried http://test.stemcellcommons.org/media/file_store/15/d4/1_tophat_accepted_hits.bai`

BAI file might be available in Galaxy: https://wiki.galaxyproject.org/Learn/GalaxyNGS101#Understanding_and_manipulating_SAM.2FBAM_datasets

- https://biostar.usegalaxy.org/p/8977/
- https://github.com/galaxyproject/tools-devteam/issues/343 (Bam indexes are handled "under the hood" within Galaxy)
- `.bai` available for download here:

![screen shot 2016-07-29 at 9 25 16 am](https://cloud.githubusercontent.com/assets/5629547/17249574/8f527fde-556e-11e6-92e6-547de8750922.png)

**Comment by scottx611x** [comment]
TODO:
- [ ] Figure out if bioblend get fetch this `.bai` file?
- [ ] Figure out if IGV session file can handle using the `.bai` if its in a different dir after utilizing another tool to generate index?

**Comment by hackdna** [comment]
"The index files must have the same file name and must reside in the same directory as the file that it indexes. For example, the index file for test-xyz.bam would be named test-xyz.bam.bai."
https://www.broadinstitute.org/igv/BAM

**Comment by scottx611x** [comment]
After implementing the automatic generation of `.bai` files  upon the import of a `bam` file in Refinery, I've run into this error while trying to launch IGV.

```
Errors were encountered loading the session: http://192.168.50.50:8000/media/file_store/0b/a1/wgEncodeUwRepliSeqK562G1AlnRep1.bam &nbs;p org.broad.igv.exceptions.DataLoadException: Read error; BinaryCodec in readmode; streamed file (filename not available)
```

Seems like this is a known issue: http://dev.list.galaxyproject.org/Galaxy-amp-IGV-display-error-td4657210.html

**Comment by scottx611x** [comment]
```
For IGV and most external displays that work on BAM and other larger seekable file formats (bigwig, bigbed, etc.) you'll need to use and configure a proxy server for your local Galaxy instance that provides byte-range request capabilities, such as nginx (http://wiki.galaxyproject.org/Admin/Config/Performance/nginx%20Proxy) or apache (http://wiki.galaxyproject.org/Admin/Config/Apache%20Proxy). 
```

**Comment by hackdna** [comment]
Yes, you need to test this functionality via Apache. Django dev server has very limited capabilities.

**Comment by scottx611x** [comment]
This is working locally in production mode with some smaller bams. I'll deploy an AWS instance and do some more exhaustive testing today

**Comment by scottx611x** [comment]
#409 is potentially related here

**Comment by scottx611x** [comment]
As of commit: 1af3cf86dfb6415d444fc2060239775bec43203b in the `scottx611x/enable_viewing_bam_in_igv` branch, files that are imported into Refinery that are of `FileType: BAM` will automatically have the proper `.bai` file generated and associated with their "parent" `FileStoreItem` and be properly deleted when their "parent" file is, but this may not be the best way to resolve this issue.

With this being said, and even with the news from the Galaxy folks today about being able to download these index files after a `.bam` has been analyzed. The issue still exists that one cannot import their Data and immediately vizualize it in IGV without either importing a tabular file and checking `Datafile permanent` or analyzing this file in Galaxy first.

@ngehlenborg has stated that it would be good to have the functionality to generate these types of files locally, and I agree that it could be helpful down the road for other types of index files, but I don't think we should waste resources computing this index right if we can get it from Galaxy and have no way of allowing the user to visualize in IGV without said file being imported first..

We could always try to generate this index whenever we see a Dataset uploaded with a file (local or remote) with the`.bam` extension, but then we need to keep in mind that this index could potentially be lying around without its "parent" file in the same directory like IGV expects if we do things this way (Could always point the IGV session file to remote bam and local index file, but then things start getting messy and we would have to handle a couple different use cases).

#### TL;DR

- Should we generate be generating these index files locally if they exist (and are now accessible) in Galaxy (Both scenarios still rely on a file_import task to be called)?
- How should we associate these index files with their "parent" files so we can properly utilize IGV (through Node model)?
- How should we go about allowing Users to view these bams in IGV (Only after file has been imported into Refinery or whenever they please)?

@ngehlenborg @hackdna .. thoughts???

**Comment by hackdna** [comment]
FileStoreItems correspond to and deal with a single file, so index files should be managed using their own FileStoreItem objects. Association of an index to the primary file should happen via a separate mechanism (separation of concerns).
Other questions don't have obvious answers and require more analysis.

**Comment by scottx611x** [comment]
@hackdna the index_file field is a OneToOne field on the FileStoreItem model in this case, so it is tied to its own FileStoreItem instance.

**Comment by hackdna** [comment]
I meant FileStoreItems should not need to be aware of other FileStoreItems. That linkage should happen on a different level of abstraction.

**Comment by ngehlenborg** [comment]
Thanks @scottx611x for writing down the details of what we have right now and where we could/should take this.

I think the best solution to handle this would be attach such auxiliary files (which in our case will be used almost exclusively for visualization purposes) at the `Node` level. For instance, a node that represents a _bam_ file would have a child node representing the _bai_ file. That child node would have a special flag to indicate that it is representing an auxiliary file, and there could be multiple such nodes that represent auxiliary files. The `Node` API would have to be extended to provide access to auxiliary nodes upon request. Since this is getting us closer to the visualization abstraction layer, we should involve @mccalluc in this conversation as well.

We still have to decide if we want to make auxiliary file nodes available as inputs for analyses and how we compute them in general.

**Comment by scottx611x** [comment]
**Steps moving forward:**

- [ ] Create Node api v2 with: get_parents(), get_children(), get_auxiliary/indexfile_node() and full_url_of_data_file field
- [ ] Have some setting to be able to chose how/when these index files get created: `['never', 'upon_import_into_refinery']` ** This setting could be `FileType` secific
- [ ] Associate node that is representative of the generated index file with its parent -> `add_child()`
- [ ] Add kwarg to api call (like we do with `/sharing/`) to be able to retrieve the normal Node api response or adding `/auxilliary_nodes/` would return nodes plus their index files
- [ ] Generate some status for the creation of these indicies and show that in the front end

**Comment by hackdna** [comment]
Looks good, except we should compute the full URLs of data files at runtime instead of storing in a field.

**Comment by scottx611x** [comment]
@hackdna: Chuck may end up needing these full_url's exposed through the Node api v2 for the new IGV integration

**Comment by ngehlenborg** [comment]
Maybe a misunderstanding about the `full_url_of_data_file` field? My support is for not storing this in the database but generating that on the fly as part of the API response. I don't think that the client should compute the full URL.

**Comment by scottx611x** [comment]
@ngehlenborg precisely, `full_url_of_data_file` will be generated whenever the new api is hit

**Comment by hackdna** [comment]
OK but overall, API should not return URLs that contain the hostname and URL scheme unless absolutely necessary for external applications.

**Comment by scottx611x** [comment]
**Finishing up:**

A new issue shouldn't be necessary for the DRF auth stuff. We can access the request.user object available 
in our viewsets and APIView methods, and should just be utilizing the `is_authenticated`, `is_authorized` methods then failing the responses appropriately.

- [ ] add test for aux_node_gen task state in Node v2 api
- [ ] remove other `get_file_extension` method and rename `get_fileextension`
- [ ] update aux_node meta message
- [ ] add all v2 apis to DRF api root
- [ ] add Django swagger yaml
- [ ] No limit on the amount of auxiliary nodes for a single node
- [ ] make sure aux Nodes get deleted properly
- [ ] Comments in NodeSerializer
- [ ] refactor `get_file_extension` in NodeSerializer (use Node model method instead)
- [ ] `create_and_associate_auxiliary_node` is "code smelly"
- [ ] `get_aux_node_generation_task_state` should return info for all aux nodes! Not just one!
- [ ] should the `uuid` field exposure for child/parent Nodes be done at the API level?
- [ ] Do I really need another fixture for the `used for visualiztion` field?
- [ ] Node.get_file_extension's docstring needs updating

**Comment by hackdna** [comment]
Looks like adding [SessionAuthentication](http://www.django-rest-framework.org/api-guide/authentication/#sessionauthentication) should be enough to enable authentication for the API endpoints?

**Comment by scottx611x** [comment]
@hackdna I think that should work just fine! I'll read up more on other options as well

**Comment by scottx611x** [comment]
After talking with @hackdna today we saw that my current implementation here lets `FileStoreItems` know about `Nodes`. We've agreed that `FIleStoreItems` are their own separate entity, and this behavior should not be allowed.

To combat this:
- it looks like the mapping of Node.file_uuid -> FileStoreItem.uuid is happening for each `Node` to be imported before `import_file` task calls (which is good, but still need to check out all usages in codebase) 
- If we have this knowledge, we should be able to run these tasks, but monitor their `state` through some polling Node model method calls utilizing `FileStoreItem.taskset_id` that we know of from each of the `import_file` calls
- We can then apply the logic of whether or not we should `generate_auxiliary_node`s (which should be moved into the Node model as well) after we detect a `SUCCESS` state from any given `taskset_id`

**Comment by hackdna** [comment]
```
Node.get_file():
    task_id = import_file()
    if viz_file:
        generate_aux_file(task_id)

@task
generate_aux_file():
    if import_file.not_ready():
        retry()
    if import_file.successfull:
        if bam:
            generate_bam_file()
        if wig:
            generate_wig_file()
```

**Comment by scottx611x** [comment]
Closed per merge of #1390

**Comment by hackdna** [comment]
Doesn't work on https://beta.stemcellcommons.org (commit 0a3d1b39).

**Comment by hackdna** [comment]
new issue #1491

**Comment by hackdna** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [IGV cannot visualize BAM files produced by RNA-seq workflows](https://trello.com/c/mL01SbVo/325-igv-cannot-visualize-bam-files-produced-by-rna-seq-workflows)
